core/src/main/java/org/infinispan/interceptors/distribution/ScatteredDistributionInterceptor.java

```diff
@@ -202,14 +207,19 @@ private Object handleWriteCommand(InvocationContext ctx, DataWriteCommand comman
                Metadata metadata = addVersion(cacheEntry.getMetadata(), nextVersion);
                cacheEntry.setMetadata(metadata);
 
+               CompletionStage<Void> stage;
                if (cmd.loadType() != DONT_LOAD) {
-                  commitSingleEntryIfNoChange(cacheEntry, rCtx, cmd);
+                  stage = commitSingleEntryIfNoChange(cacheEntry, rCtx, cmd);
                } else {
-                  commitSingleEntryIfNewer(cacheEntry, rCtx, cmd);
+                  stage = commitSingleEntryIfNewer(cacheEntry, rCtx, cmd);
                }
 
                Object returnValue = cmd.acceptVisitor(ctx, new PrimaryResponseGenerator(cacheEntry, rv));
-               return singleWriteResponse(rCtx, cmd, returnValue);
+               if (CompletionStages.isComplete(stage)) {
+                  return singleWriteResponse(rCtx, cmd, returnValue);
```

Comment:
Oh I see what you mean, because if it inlines the `addCallback` method it can possibly remove it. Sadly I would guess this method is way too big to probably get any inlining ;( It is 80 lines alone, not even counting if it inlines other stuff before.

Reply:
I will change it though.